`Type:of` does not support types without a size

I tried this code:

```rust
#![feature(type_info)]

use std::mem::type_info::Type;

fn main() {
    let type_ = const { Type::of::<str>() };
}
```

I expected this to compile successfully.

Instead, I received an error that `Type::of` doesn't support unsized types:

```
error[E0277]: the size for values of type `str` cannot be known at compilation time
  --> ~/dev/rust-lang/rust/tests/ui/reflection/size.rs:11:37
   |
LL |     let _type_ = const { Type::of::<str>() };
   |                                     ^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `str`
note: required by an implicit `Sized` bound in `Type::of`
```

I believe this occurs because `Type::of` doesn't have an `?Sized` bound. As a temporary workaround, I used `TypeId::of` first, and then accessed the info from there:

```rust
let type_ = const { TypeId::of::<str>().info() };
```

### Meta

This is on the `main` branch (commit 137716908d561bfcf0341701ec13a28326926c82). This uses the `type_info` feature (#146922), and is related to #146923.

--- a/library/core/src/mem/type_info.rs
+++ b/library/core/src/mem/type_info.rs
@@ -31,7 +31,7 @@ impl Type {
     #[unstable(feature = "type_info", issue = "146922")]
     #[rustc_const_unstable(feature = "type_info", issue = "146922")]
     // FIXME(reflection): don't require the 'static bound
-    pub const fn of<T: 'static>() -> Self {
+    pub const fn of<T: ?Sized + 'static>() -> Self {
         const { TypeId::of::<T>().info() }
     }
 }

--- a/tests/ui/reflection/dump.rs
+++ b/tests/ui/reflection/dump.rs
@@ -27,4 +27,6 @@ fn main() {
     println!("{:#?}", const { Type::of::<&Unsized>() }.kind);
     println!("{:#?}", const { Type::of::<&str>() }.kind);
     println!("{:#?}", const { Type::of::<&[u8]>() }.kind);
+    println!("{:#?}", const { Type::of::<str>() }.kind);
+    println!("{:#?}", const { Type::of::<[u8]>() }.kind);
 }

--- a/tests/ui/reflection/dump.run.stdout
+++ b/tests/ui/reflection/dump.run.stdout
@@ -21,3 +21,5 @@ Other
 Other
 Other
 Other
+Other
+Other

